api/src/main/java/org/apache/iceberg/types/IndexByName.java

```diff
@@ -23,84 +23,97 @@
 import com.google.common.collect.Lists;
 import com.google.common.collect.Maps;
 import java.util.Deque;
+import java.util.List;
 import java.util.Map;
-import java.util.concurrent.Callable;
-import java.util.function.Supplier;
 import org.apache.iceberg.Schema;
 import org.apache.iceberg.exceptions.ValidationException;
 
-public class IndexByName extends TypeUtil.CustomOrderSchemaVisitor<Map<String, Integer>> {
+public class IndexByName extends TypeUtil.SchemaVisitor<Map<String, Integer>> {
   private static final Joiner DOT = Joiner.on(".");
 
   private final Deque<String> fieldNames = Lists.newLinkedList();
   private final Map<String, Integer> nameToId = Maps.newHashMap();
 
   @Override
-  public Map<String, Integer> schema(Schema schema, Supplier<Map<String, Integer>> structResult) {
-    return structResult.get();
+  public void beforeField(Types.NestedField field) {
+    fieldNames.push(field.name());
   }
 
   @Override
-  public Map<String, Integer> struct(Types.StructType struct, Iterable<Map<String, Integer>> fieldResults) {
-    // iterate through the fields to update the index for each one, use size to avoid errorprone failure
-    Lists.newArrayList(fieldResults).size();
-    return nameToId;
+  public void afterField(Types.NestedField field) {
+    fieldNames.pop();
   }
 
   @Override
-  public Map<String, Integer> field(Types.NestedField field, Supplier<Map<String, Integer>> fieldResult) {
-    withName(field.name(), fieldResult::get);
-    addField(field.name(), field.fieldId());
-    return null;
+  public void beforeListElement(Types.NestedField elementField) {
+    if (!elementField.type().isStructType()) {
```

Comment:
Do we need to add some description of why return when `isStructType()` is true? IIRC, it causes duplicated names.

Reply:
Yeah, I'll add a comment. The reason is to make the names more natural. If you have a list called `locations` that has `latitude` and `longitude` fields, it would be strange to reference `locations.element.latitude` instead of `locations.latitude`. But if the element is not a struct, that doesn't apply.